discord: Fix action timeout, add in game time elapsed option

Matthew-nop opened the pull request:
Fixed the discord plugin's action timeout to match expected behavior and clear the activity from presence. Previously if there was only an event removed (such as training herblore), then the presence was not updated. Updating with the current game state fixes this.

I've also added an option to show total time in game rather than time elapsed performing the current activity. This option is the default.
Closes #12438

deathbeam commented:
@Matthew-nop Can you recheck? I cleaned up this a bit and also decided to go with the correct updating instead of force triggering of events. Based on my testing its working fine (note: the total time is now reset after you login/logout, so only total in-game time and total in-menu time is tracked as imo that makes more sense).

Matthew-nop commented:
@deathbeam thanks for the work on this PR.
I've changed/fixed a few things I found while testing:
- The elapsed time would reset on teleporting due to `checkForAreaEvent()` triggering an `IN_GAME` event. I changed `IN_GAME` to have `shouldRestart` equal false to remedy this.
- The state would not update when moving from a known region (e.g. Al Kharid) to a generic unknown region (In Game). I made `triggerEvent` clear all clearable events on a `IN_GAME` event to fix this. 
- I changed `shouldClear` on `IN_MENU` to true so an `IN_GAME` event will clear it along with all other areaEvents. 
If there's anything you'd like changed please let me know.

CoopJax commented:
Oh hey, isn't this what I was asking for?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordState.java:
```diff
@@ -115,32 +117,51 @@ void refresh()
+		// An IN_GAME (or IN_MENU) event should remove all other areas from events
+		// but since events is fully cleared every time before IN_MENU is added,
+		// we will never need to run this on an IN_MENU event trigger
+		if (event.getType().isShouldClear() || event.getType() == DiscordGameEventType.IN_GAME)
```
Imo this behaviour should be controlled by enum. Basically what you want here is flag to mark IN_GAME event to be "sticky" or something, e.g it should not be removed when shouldClear event triggers. Or add shouldBeCleared and shouldClear flags where shouldClear will clear other events that need to be cleared and shouldBeCleared will determine if the event will be cleared.

So in this case you would end up with IN_GAME event with shouldBeCleared: false, shouldClear: true

deathbeam commented:
I tried to incorporate what I mentioned in my comment above in 8ccea05. What i did is:

1. I added `root` flag to game events to avoid checking specifically for IN_GAME or IN_MENU when determining "root" event what should be used for total time tracking
2. I added `shouldBeCleared` flag that determines if event should be cleared when processed
3. I changed original `shouldClear` flag to just determine if event should clear *other* events when triggered

I also added documentation for each flag so hopefully this will be less confusing.

Matthew-nop commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordState.java:
```diff
@@ -209,10 +206,10 @@ private void updatePresenceWithLatestEvent()
+					.filter(e -> e.getType().isRoot())
 					.sorted((a, b) -> b.getUpdated().compareTo(a.getUpdated()))
 					.map(EventWithTime::getStart)
 					.findFirst()
```
Assuming there's only 1 root event in events list at any given time, this could be simplified.

deathbeam replied:
Well in-menu theoretically can be active at same time as in-game timer

Matthew-nop replied:
Yeah, this is just to clarify which section of code I was referring to in the above post. Although both can be in the events list, is it possible for the player to actually be both in-menu and in-game at the same time?

Matthew-nop reviewed (commented):
With the current implementation, it's possible for `events` to contain both an `IN_GAME` and `IN_MENU` event which seems odd given that the player can't concurrently be both in menu and in game.

Would it be a good idea to enforce there being strictly 1 root event in the events list? I'm not sure if it makes sense that there is more than 1 root in the list, and this way we could simplify the code in the switch statement for total time tracking. When processing a new and unique root event we could clear all other root events.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -451,10 +451,32 @@
+	 */
 	private boolean shouldRestart;
 
+	/**
+	 * Determines if event should be cleared when processed
+	 */
+	private boolean shouldBeCleared = true;
```
processed indicates that its cleared when that happens (as events are processed either when triggered or in the periodic timeout check). and i think shouldBeCleared is more clear, idk
